File: q2/JAVA/q2.java
public class Solution {
    public String replaceSpace(StringBuffer str) {
        int blankCount = 0;
        for (int i = 0; i < str.length(); i++) {
            if (str.charAt(i) == ' ') {
                blankCount++;
            }
        }
        int preIndex = str.length() - 1;
        int newLength = str.length() + 2 * blankCount;
        str.setLength(newLength);
        int newIndex = str.length() - 1;

        while (preIndex >= 0 && newIndex >= 0) {
            if (str.charAt(preIndex) == ' ') {
                str.setCharAt(newIndex--, '0');
                str.setCharAt(newIndex--, '2');
                str.setCharAt(newIndex--, '%');
            } else {
                str.setCharAt(newIndex--, str.charAt(preIndex));
            }
            preIndex--;
        }
        return str.toString();
    }
}

File: q2/C++/q2.cpp
class Solution {
public:
    void replaceSpace(char *str,int length) {
        if (length < 0 && str == nullptr)
            return;
        int oriLength = 0;
        int blankCount = 0;
        
        int i = 0;
        while (str[i] != '\0') {
            oriLength++;
            if (str[i] == ' ')
                blankCount++;
            i++;
        }
        int newLength = oriLength + 2 * blankCount;
        if (length < newLength + 1)
            return;
        str[newLength] = '\0';
        i = newLength - 1;
        int j = oriLength - 1;
        while (i >= 0 && j >= 0) {
            if (str[j] == ' ') {
                str[i--] = '0';
                str[i--] = '2';
                str[i--] = '%';
            } else {
                str[i--] = str[j];
            }
            j--;
        }
    }
};
